marklogic-spark-connector/src/main/java/com/marklogic/spark/writer/DocBuilder.java

```diff
@@ -138,22 +171,57 @@ private DocumentWriteOperation buildMainDocument(DocumentInputs inputs) {
         final String graph = inputs.getGraph();
         final DocumentMetadataHandle metadataFromRow = inputs.getInitialMetadata();
 
+        BytesHandle content = (BytesHandle) inputs.getContent();
+        byte[] classificationResponse = inputs.getClassificationResponse();
+        if (classificationResponse != null && inputs.getExtractedText() == null && Format.XML.equals(content.getFormat())) {
+            content = appendDocumentClassificationToXmlContent(content.get(), inputs.getInitialUri(), classificationResponse);
+        }
+
         DocumentMetadataHandle sourceMetadata = metadataFromRow;
         if (sourceMetadata != null) {
             // If the row contains metadata, use it, but first override it based on the metadata specified by user options.
             overrideMetadataFromRowWithMetadataFromOptions(sourceMetadata);
             if (graph != null) {
                 sourceMetadata.getCollections().add(graph);
             }
-            return new DocumentWriteOperationImpl(sourceUri, sourceMetadata, inputs.getContent());
+            return new DocumentWriteOperationImpl(sourceUri, sourceMetadata, content);
         }
         // If the row doesn't contain metadata, use the metadata specified by user options. We need to be careful
         // not to modify that object though, as it will be reused on subsequent calls.
         sourceMetadata = metadataFromOptions;
         if (graph != null && !sourceMetadata.getCollections().contains(graph)) {
             sourceMetadata = newMetadataWithGraph(graph);
         }
-        return new DocumentWriteOperationImpl(sourceUri, sourceMetadata, inputs.getContent());
+        return new DocumentWriteOperationImpl(sourceUri, sourceMetadata, content);
+    }
+
+    public BytesHandle appendDocumentClassificationToXmlContent(byte[] originalContent, String uri, byte[] classificationResponse) {
+        if (documentBuilderFactory == null) {
+            documentBuilderFactory = DocumentBuilderFactory.newInstance();
+        }
+        try {
+            DocumentBuilder builder = documentBuilderFactory.newDocumentBuilder();
+            Document originalDoc = builder.parse(new ByteArrayInputStream(originalContent));
+
+            Document responseDoc = builder.parse(new ByteArrayInputStream(classificationResponse));
+            NodeList structuredDocumentNodeChildNodes = responseDoc.getElementsByTagName(CLASSIFICATION_MAIN_ELEMENT).item(0).getChildNodes();
+            Node classificationNode = originalDoc.createElementNS(DEFAULT_XML_NAMESPACE,"classification");
+            for (int i = 0; i < structuredDocumentNodeChildNodes.getLength(); i++) {
+                Node importedChildNode = originalDoc.importNode(structuredDocumentNodeChildNodes.item(i),true);
+                classificationNode.appendChild(importedChildNode);
+            }
+            originalDoc.getFirstChild().appendChild(classificationNode);
+
+            return new BytesHandle(convertDomToByteArray(originalDoc));
```

Comment:
Don't need this conversion - just return a `DOMHandle(originalDoc)`. The method can return an `AbstractWriteHandle` as that's all that `DocumentWriteOperationImpl` needs. Can then remove all the Transformer stuff.

Reply:
Ugh, sorry.  I remembered this as I was falling asleep last night.